sympy/polys/numberfields/subfield.py

```diff
@@ -136,23 +141,27 @@ def field_isomorphism_pslq(a, b):
 def field_isomorphism_factor(a, b):
     """Construct field isomorphism via factorization. """
     _, factors = factor_list(a.minpoly, extension=b)
-
+    A = None
     for f, _ in factors:
         if f.degree() == 1:
-            coeffs = f.rep.TC().to_sympy_list()
+            # Any linear factor f(x) represents _some conjugate of_ a in QQ(b).
+            # We want to know whether this linear factor represents a itself.
+            # Let f = x - c
+            c = -f.rep.TC()
+            # Write c as polynomial in b
+            coeffs = c.to_sympy_list()
             d, terms = len(coeffs) - 1, []
-
             for i, coeff in enumerate(coeffs):
                 terms.append(coeff*b.root**(d - i))
-
             root = Add(*terms)
-
-            if (a.root - root).evalf(chop=True) == 0:
+            # Check whether we got the number a
+            if A is None:
+                A = a.root.evalf(chop=True)
+            if root.evalf(chop=True) == A:
```

Comment:
Checking `a.evalf() == b.evalf()` is not equivalent to checking `(a - b).evalf()` if the numbers differ by an amount smaller then the default precision of `evalf`:
```python
In [40]: a = 1 + Rational(1, 10**20)

In [41]: b = S(1)

In [42]: a.evalf() == b.evalf()
Out[42]: True

In [43]: (a - b).evalf()
Out[43]: 1.00000000000000e-20
```

Reply:
Maybe that's not a problem here though. We know that `a` and `b` are nonzero. Are they guaranteed to differ in sign or leading digit?

Reply:
Not necessarily sign or leading digit, but I am thinking that an isolating interval approach could be the way to go.

Reply:
Yes, that makes sense. What exactly does "conjugate" mean here? Does it mean that they have the same minimal polynomial?

Reply:
Yup, exactly that.

So, yeah, for each linear factor, we've found _some_ root of the minimal polynomial of `a`, and we just want to know: is this `a` itself? It does sound like a job for isolating intervals.

And I guess computing adequate isolating intervals can be faster than working with evalf.

Reply:
The root isolation code can definitely be improved so that refinement runs faster. The problem with evalf is that you need to isolate the root very closely but actually with a bounding interval if you know that `a` is a root of `p` then a bigger bounding interval is better than a smaller one:
```python
In [1]: r = RootOf(x**4 + x**3 + x**2 + x + 1, 0)

In [2]: r._get_interval()
Out[2]: (-1, -1/2) x (-1, -1/2)

In [3]: r.evalf()  # This is the slow part
Out[3]: -0.809016994374947 - 0.587785252292473⋅ⅈ

In [4]: r._get_interval()
Out[4]: (-932733090407273441/1152921504606846976, -1865466180814546881/2305843009213693952) x (-677670257458753271/1152921504606846976, -1355340514917506541/2305843009213693952)
```

Reply:
Question about `evalf`: Is there a good way to get the desired number of decimal places strictly to the right of the decimal point, without computing a log?

Example:

Consider `a = 1000000 + sqrt(2)`, and suppose I want 10 places to the right of the decimal point.
Obviously this doesn't work:
```python
>>> a.evalf(10)
1000001.414
```
I can do it this way:
```python
>>> a.evalf(10 + 1 + int(log(a, 10)))
1000001.4142135624
```
but I'm wondering if there might be some ready-made solution for this.

Reply:
I don't think that there is a ready-made solution although if you want this only for RootOf then there is probably a way to specify how small you want the bounding intervals to be.

Why exactly is this needed? There might be a different way to do this.

Reply:
Well, here's what I'm trying to do: Say I have a `Poly` `f` with root isolating intervals `I1`, `I2`, ..., `In`.
I have two `Expr`s `r` and `s`, and I know `f(r) == 0 == f(s)`. I want to determine whether `r == s`.

But I don't want to compute `(r - s).evalf()`, because it's very slow. (#15743)

My idea is to let `D` be the largest denominator occurring in any endpoint of any of the intervals `Ik`.
Then `abs(r - s) < 1/D` would suffice to conclude that `r == s`.

But to check the inequality requires use of `evalf`. To what precision? My idea is to detemine `n` large
enough to guarantee that `abs(r - r.evalf(n)) < eps = 1/(3*D)`. (And likewise for `s`.)

Then `abs(r.evalf(n) - s.evalf(n)) < eps` will suffice to conclude `r == s`.

So I was thinking of giving `Poly` a method like this:

```python
    def root_comparison_tools(f):
        roots = f.all_roots(radicals=False)
        ivals = [r._get_interval() for r in roots]
        # Let D0 be max denom among all interval endpoints.
        D0 = int(max(i.max_denom for i in ivals))
        # We need 3 times that.
        D1 = 3*D0
        # Pick n large enough to guarantee 10^n > D1.
        n = math.ceil(D1.bit_length() / 3.3) + 2
        # Make a custom evalf function that guarantees rounding error < 10^(-n).
        evalf_ = lambda x: x.evalf(n + 1 + math.ceil(int(x).bit_length() / 3.3))
        # Define a threshold that guarantees two roots of f are the same.
        eps = S(1) / D1
        return eps, evalf_
```

and then the problem could be solved like this:

```python
eps, evalf_ = f.root_compartison_tools()
if abs(evalf_(r) - evalf_(s)) < eps:
    # roots r and s are the same...
else:
    # they're different roots...

```

Reply:
So `(r - s).evalf()` is slow if `r` and `s` are equal because it ends up computing `r` and `s` to high precision in an attempt to find the significant digits of `r - s`. What's faster is computing the significant digits of something that is *nonzero*.

For each interval `Ik` (including the correct one) we expect that `abs(r - Ik.mid) - Di` does not equal zero and we can compute that to find whether it is positive or not e.g.:
```python
In [51]: p = x**4 + x + 2

In [52]: rs = [RootOf(p, n) for n in range(degree(p))]

In [53]: r1, r2, r3, r4 = roots(p)

In [54]: intervals = [r._get_interval() for r in rs]

In [55]: intervals[0]
Out[55]: (-1, -1/2) x (-1, -1/2)

In [56]: intervals[0].center
Out[56]: (mpq(-3,4), mpq(-3,4))

In [57]: intervals[0].dx
Out[57]: mpq(1,2)

In [58]: intervals[0].dy
Out[58]: mpq(1,2)

In [59]: r1.evalf()
Out[59]: -0.849848427793055 - 0.6542723031575⋅ⅈ

In [60]: def centre(i):
    ...:     a, b = i.center
    ...:     return a + b*I
    ...: 

In [61]: def radius(i):
    ...:     return i.dx + i.dy
    ...: 

In [62]: centre(intervals[0])
Out[62]: 
  3   3⋅ⅈ
- ─ - ───
  4    4 

In [63]: radius(intervals[0])
Out[63]: mpq(1,1)

In [64]: (abs(r1 - centre(intervals[0])) - radius(intervals[0])).evalf(2)
Out[64]: -0.86

In [65]: (abs(r1 - centre(intervals[1])) - radius(intervals[1])).evalf(2)
Out[65]: 0.41

In [66]: (abs(r1 - centre(intervals[2])) - radius(intervals[2])).evalf(2)
Out[66]: 0.71

In [67]: (abs(r1 - centre(intervals[3])) - radius(intervals[3])).evalf(2)
Out[67]: 1.5
```
Here `Out[64]` through to `Out[67]` proves that `r1` is the root bounded by `intervals[0]`. If we end up having multiple negatives here then it means we need to shrink the intervals and try again. This is analogous to the `choose_factor` method in minpoly: we prove a factor is the minpoly for `a` by proving that every other factor is *not* the minpoly. This is done by numerically showing that something is *not* zero which can be done reliably (unlike proving that something is zero).

Note that we don't actually need to do the `evalf(2)` explicitly here as it is precisely what `a > b` does for Expr:
```python
In [69]: abs(r1 - centre(intervals[3])) > radius(intervals[3])
Out[69]: True
```

Combining this for `r` and `s` an algorithm could be:
1. For each interval compute `(r - c)**2 < d**2`
2. Also compute `(s - c)**2 < d**2`
3. If either of `r` or `s` is definitely not in this interval discard the interval.
4. If we have one remaining interval then `r = s`.
5. If we have no remaining intervals then `r != s`.
6. If there are multiple intervals remaining, shrink them and go back to step 1.

That is:
```python
while True:
    intervals_keep = []
    for interval in intervals:
         c = centre(interval)
         d = radius(interval)
         # Should be a way to compute abs(z)**2 without intermediate square roots
         if abs(r - c)**2 < d**2 and abs(s - c)**2 < d**2:
             intervals_keep.append(interval)
    if not intervals_keep:
        return False
    elif len(intervals_keep) == 1:
        return True
    else:
        intervals = [shrink_interval(interval) for interval in intervals_keep]
```

It might seem inefficient to recalculate the `evalf` for each root but at least in the case of RootOf the work done by one `evalf` is carried over to the next. It would also be possible to wrap other kinds of expression with something that caches the results of `_eval_evalf`.

We know that `abs(r - Ik.mid) - Di` will not be zero for any real root because that would imply that `r` was rational. It might potentially be zero though for a Gaussian rational like `2 + 3*I` but it's unlikely.

Reply:
Well, this is pretty cool. : )  I like the algorithm, and I like the principle that it's easy to check when something is _nonzero_, so work by process of elimination.

The note at the end about the outside possibility of an `evalf(0)` still happening is an interesting point. What if we used a transcendental radius, like `pi*(dx + dy)`. Then it's definitely different from `abs(r - c)` which is algebraic.

Reply:
Yes, that makes sense. It could be `pi/3*(dx + dy)` to make the bound reasonably tight.

Reply:
Or in fact we want `(dx + dy)/2` so `pi/6` should work.

Reply:
Sounds good. I am also still interested to see whether the algorithm I described could work as well. I added a note correcting the way the denominator `D` is chosen.

Maybe it's worth coding them both up. Even if only one is used, I think they both contain valuable ideas that could be worth revisiting at some point.

Meanwhile, maybe the comprehensive move over to ball-radius arithmetic you mentioned [before](https://github.com/sympy/sympy/issues/15743#issuecomment-985706352) will eventually make it all irrelevant?

It does seem that a fundamental problem with `evalf()` is that it wants to give you so many digits of precision, when all you may really want is so much accuracy, i.e. to bound the rounding error by `10^(-n)`.

Reply:
The algorithm I described above is quite similar to what `minpoly` does. Is it a bad idea to just compute `minpoly(r - s)`?

Reply:
Oh, I suppose that had to `(r - s).evalf()` as well as all the factoring.

Reply:
The minimum distance between roots of a polynomial `p` can be estimated as follows. The roots of the resultant of `p(x)` and `p(x + y)` are the differences `x - x'` for all pairs of roots `(x, x')` of `p`.
```
In [1]: p = x**4 + x + 2                                                        

In [2]: q = p.subs(x, x + y)                                                    

In [3]: r = resultant(p, q, x)                                                  

In [4]: r                                                                       
Out[4]: 
 16       12       10        8        6         4
y   + 16⋅y   + 26⋅y   - 448⋅y  + 432⋅y  + 2021⋅y
```
After removing the zero roots, a lower bound for `|1/y|` can be found by applying e.g. [Cauchy's bound](https://en.wikipedia.org/wiki/Geometrical_properties_of_polynomial_roots#Lagrange's_and_Cauchy's_bounds) to the reversed polynomial. In the above case we get `|1/y| <= 1 + 448/2021`. The result can be sharpened by noting that the resultant is an even polynomial (for each `x - x'`, also `x' - x` is a root) hence a polynomial in `y**2` so that Cauchy's bound is also valid for `|1/y**2|`.

Reply:
Nice! So maybe the `Poly` class can have a method roughly like this:

```python
    def root_separation_lower_bound(f):
        """
        Compute a lower bound on the minimum distance between any two
        roots of this polynomial.
        """
        if f.is_multivariate:
            raise MultivariatePolynomialError(
                "must be a univariate polynomial")

        from sympy.functions.elementary.miscellaneous import sqrt

        x = f.gen
        y = Dummy('y')
        g = f.subs(x, x + y)
        h = Poly(resultant(f, g, x))
        # If f has roots ri, the roots of h are all the ri - rj.

        coeffs = h.rep.rep
        while coeffs[-1] == 0:
            coeffs.pop()
        # Use the Cauchy bound
        d = coeffs.pop()
        M = 1 + max(abs(n / d) for n in coeffs)
        # Since h is an even func., can improve bound with sqrt.
        return 1/sqrt(M)
```

Reply:
Yes, and maybe there could also be a method just called `roots_bound` that just gives root bounds in general. Using `sqrt` here is tricky. We ideally want the bounds to be rational numbers. It might be better to have something that returns the square of the bound(s).

Also `rep.rep` should not be manipulated directly as it is a private internal attribute of the immutable `Poly`. There is `all_coeffs` to get the coefficients as `Expr`.

Also any methods like this should check the domain and whether the `Poly` is univariate.

Reply:
Maybe `isqrt` could be used. Something like `Rational(isqrt(M.p*M.q) + 1, M.q)`.

Reply:
Finally was able to get back to this...

On making the bounds rational: I've written it so that the bounds will be rational if the polynomial's coefficients are. Do you think this is good enough?

Reply:
@jksuom Is it all right if I set you as first author of 9c91d52a6f? These are your ideas.

Reply:
It is all right, though I wasn't expecting that.